pkg/R/RDD.R

```diff
@@ -1286,48 +1311,44 @@ setMethod("partitionBy",
-
-            packageNamesArr <- .jarray(serialize(.sparkREnv$.packages,
+            serializedHashFuncBytes <- serialize(as.character(substitute(partitionFunc)),
                                                  connection = NULL,
-                                                 ascii = TRUE))
-            refs <- lapply(ls(.broadcastNames), function(name) {
-                           get(name, .broadcastNames) })
-            broadcastArr <- .jarray(refs,
-                                    "org/apache/spark/broadcast/Broadcast")
-            jrdd <- getJRDD(rdd)
+                                                 ascii = TRUE)
 
+            packageNamesArr <- serialize(.sparkREnv$.packages,
+                                         connection = NULL,
+                                         ascii = TRUE)
+            broadcastArr <- lapply(ls(.broadcastNames), function(name) {
+                                   get(name, .broadcastNames) })
+            jrdd <- getJRDD(rdd)
 
             # We create a PairwiseRRDD that extends RDD[(Array[Byte],
             # Array[Byte])], where the key is the hashed split, the value is
             # the content (key-val pairs).
-            pairwiseRRDD <- new(J("edu.berkeley.cs.amplab.sparkr.PairwiseRRDD"),
-                                jrdd$rdd(),
-                                as.integer(numPartitions),
-                                serializedHashFuncBytes,
-                                rdd@env$serialized,
-                                depsBinArr,
-                                packageNamesArr,
-                                as.character(.sparkREnv$libname),
-                                broadcastArr,
-                                jrdd$classTag())
+            pairwiseRRDD <- newJObject("edu.berkeley.cs.amplab.sparkr.PairwiseRRDD",
+                                       callJMethod(jrdd, "rdd"),
+                                       as.integer(numPartitions),
+                                       serializedHashFuncBytes,
+                                       rdd@env$serialized,
+                                       depsBinArr,
+                                       packageNamesArr,
+                                       as.character(.sparkREnv$libname),
+                                       broadcastArr,
+                                       callJMethod(jrdd, "classTag"))
 
             # Create a corresponding partitioner.
-            rPartitioner <- new(J("org.apache.spark.HashPartitioner"),
-                                as.integer(numPartitions))
+            rPartitioner <- newJObject("org.apache.spark.HashPartitioner",
+                                       as.integer(numPartitions))
 
             # Call partitionBy on the obtained PairwiseRDD.
-            javaPairRDD <- pairwiseRRDD$asJavaPairRDD()$partitionBy(rPartitioner)
+            javaPairRDD <- callJMethod(pairwiseRRDD, "asJavaPairRDD")
```

Comment:
We can only call `partitionBy` on something that extends a JavaPairRDD (or has PairRDDFunctions implicit functions but AFAIK that only works in Scala).